app/models/custom_button.rb

```diff
@@ -88,11 +88,31 @@ def applies_to=(other)
     end
   end
 
-  def invoke(target)
+  def invoke(target, source = nil)
     args = resource_action.automate_queue_hash(target, {}, User.current_user)
+
+    publish_event(source, target, args)
     MiqQueue.put(queue_opts(target, args))
   end
 
+  def publish_event(source, target, args)
+    CustomEvent.create(
+      :event_type  => 'custom_button_event',
+      :message     => 'Custom button launched',
+      :source      => source,
+      :type        => 'CustomEvent',
+      :target_id   => target.id,
+      :target_type => target.type,
+      :user_id     => args[:user_id],
+      :group_id    => args[:miq_group_id],
+      :tenant_id   => args[:tenant_id],
+      :full_data   => {
+        :args                 => args,
+        :automate_entry_point => resource_action.ae_path
+      }
+    )
```

Comment:
The `CustomEvent` model and its parent `EventStream` is a little bit unclear for me. In the UI we're using `EventMixin#has_events?` for enabling/disabling timeline buttons for a given entity. This invokes an `event_where_clause` on the entity that doesn't always operate with the `target`. For example for providers, it's checking the [`ems_id`](https://github.com/ManageIQ/manageiq/blob/master/app/models/ext_management_system.rb#L549) that this method doesn't even set. For VMs it's a little [trickier](https://github.com/ManageIQ/manageiq/blob/master/app/models/vm_or_template.rb#L1383-L1390), depending on the event type, it checks the `vm_or_template_id` or the `target`.

So this is a little confusing ... the ugly solution would be to edit all the `event_where_clause` methods, but I really hope :pray: :pray: there's a better solution for this.

Reply:
ping @bronaghs as @tumido is on PTO and you will probably know who should I ask :wink:

Reply:
oh, probably @gmcculloug

Reply:
@dclarizio pinging you here as you asked me to do timelines :wink: